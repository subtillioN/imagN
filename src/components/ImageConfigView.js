import React from 'react';
import { BaseView } from '../mvi/view';
import styles from '../styles/base.module.css';

export class ImageConfigView extends BaseView {
  constructor(sources) {
    super(sources);
    this.initializeStreams();
    this.state = {
      config: {},
      parameters: {}
    };
  }

  initializeStreams() {
    this.config$ = this.sources.state.config$ || {};
    this.parameters$ = this.sources.state.parameters$ || {};

    // Subscribe to stream updates
    this.config$.subscribe(config => {
      this.setState({ config });
    });

    this.parameters$.subscribe(parameters => {
      this.setState({ parameters });
    });
  }

  renderParameters(state) {
    return (
      <div className={styles['parameters-section']}>
        <div className={styles['parameter-group']}>
          <label>Image Size</label>
          <select className={styles['size-select']}>
            <option value="512x512">512x512</option>
            <option value="1024x1024">1024x1024</option>
          </select>
        </div>
        <div className={styles['parameter-group']}>
          <label>Style</label>
          <select className={styles['style-select']}>
            <option value="realistic">Realistic</option>
            <option value="artistic">Artistic</option>
            <option value="abstract">Abstract</option>
          </select>
        </div>
        <div className={styles['parameter-group']}>
          <label>Prompt</label>
          <input
            className={styles['prompt-input']}
            type="text"
            placeholder="Describe your image..."
          />
        </div>
      </div>
    );
  }

  renderControls(state) {
    return (
      <div className={styles['controls-section']}>
        <button className={styles['generate-btn']}>Generate Image</button>
        <button className={styles['reset-btn']}>Reset</button>
      </div>
    );
  }

  renderPreview(state) {
    return (
      <div className={styles['preview-section']}>
        <div className={styles['preview-container']}>
          {/* Preview will be implemented here */}
        </div>
      </div>
    );
  }

  view(state) {
    return (
      <div className={styles['image-config-container']}>
        <div className={styles['config-header']}>Image Configuration</div>
        {this.renderParameters(state)}
        {this.renderPreview(state)}
        {this.renderControls(state)}
      </div>
    );
  }
}