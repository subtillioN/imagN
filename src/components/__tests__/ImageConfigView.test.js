import { ImageConfigView } from '../ImageConfigView';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { makeSubject } from 'callbag-subject';

describe('ImageConfigView Component', () => {
  const createMockStream = () => {
    const subject = makeSubject();
    return {
      subscribe: (callback) => {
        subject(0, callback);
        return () => subject(2);
      }
    };
  };

  const createSources = () => ({
    state: {
      config$: createMockStream(),
      parameters$: createMockStream()
    }
  });

  it('should render correctly and match snapshot', () => {
    const sources = createSources();
    const imageConfigView = new ImageConfigView(sources);
    const { container } = render(imageConfigView.view({}));
    
    expect(container).toMatchSnapshot();
  });

  it('should render all configuration sections', () => {
    const sources = createSources();
    const imageConfigView = new ImageConfigView(sources);
    const { container } = render(imageConfigView.view({}));

    expect(container.querySelector('.image-config-container')).toBeInTheDocument();
    expect(container.querySelector('.parameters-section')).toBeInTheDocument();
    expect(container.querySelector('.preview-section')).toBeInTheDocument();
    expect(container.querySelector('.controls-section')).toBeInTheDocument();
  });

  it('should render all parameter inputs', () => {
    const sources = createSources();
    const imageConfigView = new ImageConfigView(sources);
    const { container } = render(imageConfigView.view({}));

    expect(container.querySelector('.size-select')).toBeInTheDocument();
    expect(container.querySelector('.style-select')).toBeInTheDocument();
    expect(container.querySelector('.prompt-input')).toBeInTheDocument();
    expect(container.querySelector('.generate-btn')).toBeInTheDocument();
    expect(container.querySelector('.reset-btn')).toBeInTheDocument();
  });

  it('should update state when streams emit new values', () => {
    const sources = createSources();
    const imageConfigView = new ImageConfigView(sources);
    const { container } = render(imageConfigView.view({}));

    // Simulate stream updates
    sources.state.config$.subscribe(callback => callback({ theme: 'dark' }));
    sources.state.parameters$.subscribe(callback => callback({ size: '1024x1024' }));

    expect(imageConfigView.state.config).toEqual({ theme: 'dark' });
    expect(imageConfigView.state.parameters).toEqual({ size: '1024x1024' });
  });
});