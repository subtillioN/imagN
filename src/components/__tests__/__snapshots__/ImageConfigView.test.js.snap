// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageConfigView Component should render correctly and match snapshot 1`] = `
<div>
  <div
    class="image-config-container"
  >
    <div
      class="config-header"
    >
      Image Configuration
    </div>
    <div
      class="parameters-section"
    >
      <div
        class="parameter-group"
      >
        <label>
          Image Size
        </label>
        <select
          class="size-select"
        >
          <option
            value="512x512"
          >
            512x512
          </option>
          <option
            value="1024x1024"
          >
            1024x1024
          </option>
        </select>
      </div>
      <div
        class="parameter-group"
      >
        <label>
          Style
        </label>
        <select
          class="style-select"
        >
          <option
            value="realistic"
          >
            Realistic
          </option>
          <option
            value="artistic"
          >
            Artistic
          </option>
          <option
            value="abstract"
          >
            Abstract
          </option>
        </select>
      </div>
      <div
        class="parameter-group"
      >
        <label>
          Prompt
        </label>
        <input
          class="prompt-input"
          placeholder="Describe your image..."
          type="text"
        />
      </div>
    </div>
    <div
      class="preview-section"
    >
      <div
        class="preview-container"
      />
    </div>
    <div
      class="controls-section"
    >
      <button
        class="generate-btn"
      >
        Generate Image
      </button>
      <button
        class="reset-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
